crates/runtime/src/execution_plan/fallback_scan.rs

```diff
@@ -0,0 +1,217 @@
+        "FallbackScanExec"
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+
+    fn properties(&self) -> &PlanProperties {
+        &self.properties
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![Arc::clone(&self.input)]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        if children.len() == 1 {
+            Ok(Arc::new(FallbackScanExec::new(
+                Arc::clone(&children[0]),
+                Arc::clone(&self.fallback_table_provider),
+                self.fallback_scan_params.clone(),
+            )))
+        } else {
+            Err(DataFusionError::Execution(
+                "FallbackScanExec expects exactly one input".to_string(),
+            ))
+        }
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        tracing::trace!("Executing FallbackScanExec: partition={}", partition);
+        if partition > 0 {
+            return Err(DataFusionError::Execution(format!(
+                "FallbackScanExec only supports 1 partitions, but partition {partition} was requested",
+            )));
+        }
+
+        let mut input_stream = self.input.execute(0, Arc::clone(&context))?;
+        let schema = input_stream.schema();
+        let scan_params = self.fallback_scan_params.clone();
+        let fallback_provider = Arc::clone(&self.fallback_table_provider);
+
+        let potentially_fallback_stream = stream::once(async move {
+            let context = Arc::clone(&context);
+            let schema = input_stream.schema();
+            // If the input_stream returns a value - then we don't need to fallback. Piece back together the input_stream.
+            if let Some(input) = input_stream.next().await {
+                tracing::trace!("FallbackScanExec input_stream.next() returned Some()");
+                match &input {
+                    Ok(batch) => {
+                        tracing::trace!(
+                            "FallbackScanExec input_stream.next() is Ok(): num_rows: {}",
+                            batch.num_rows()
+                        );
+                    }
+                    Err(e) => {
+                        tracing::trace!("FallbackScanExec input_stream.next() is Err(): {e}");
+                    }
+                }
+                // Add this input back to the stream
+                let input_once = stream::once(async move { input });
+                let stream_adapter =
+                    RecordBatchStreamAdapter::new(schema, input_once.chain(input_stream));
+                Box::pin(stream_adapter) as SendableRecordBatchStream
+            } else {
+                tracing::trace!("FallbackScanExec input_stream.next() returned None");
```

Comment:
How the caller of this knows when fallback happens, and provide info to the user?

Reply:
We could potentially use the built-in `metrics` function on the ExecutionPlan for this:

```rust
/// Return a snapshot of the set of [`Metric`]s for this
    /// [`ExecutionPlan`]. If no `Metric`s are available, return None.
    ///
    /// While the values of the metrics in the returned
    /// [`MetricsSet`]s may change as execution progresses, the
    /// specific metrics will not.
    ///
    /// Once `self.execute()` has returned (technically the future is
    /// resolved) for all available partitions, the set of metrics
    /// should be complete. If this function is called prior to
    /// `execute()` new metrics may appear in subsequent calls.
    fn metrics(&self) -> Option<MetricsSet> {
        None
    }
```